### ISSUE TITLE
ArrayPool Create with max array length < 1040 bytes

### ISSUE THREAD

**Issue by stefan2410** [dotnet/runtime#51919]
ArrayPool<byte>.Create  with  max array length  < 1040 bytes, returns always  1040.

I understand that the allocation of smaller buffers probably will be less efficient than the GC allocation, but we intend to create a cache of such small buffers, so the allocation's cost is not problem.  But we waste more memory if  we allocate, at once, a million buffers of 640 bytes and ArrayPool returns  1040 minimum length.

**Comment by gfoidl** [comment]
> ArrayPool.Create with max array length < 1040 bytes, returns always 1040.

How do you create the array pool? With default config or with the shared pool the minimum size is 16.

> the allocation's cost is not problem

Allocations are very fast and the cost is minor. You have to pay the cost at cleaning up the garbage, but the GC is pretty good.
Pooling (renting / returning and the pool-management) also has a cost. A pool is kind of a custom allocator.
So please run some benchmarks to verify that pooling actually proves as the better option. It can harm performance too (pooled objects will go to Gen-2, which means more work there amongst others).

Maybe (after profiling) something like a [slab memory-pool](https://github.com/dotnet/aspnetcore/blob/1453c5bd2cf886f6846d2cc00b8106dbd0fdbd53/src/Shared/Buffers.MemoryPool/PinnedBlockMemoryPool.cs#L13) suits your needs?
Basically this allocates a larger block, then slices it to needed sizes and returns that memory. At return these memories are pooled.
This is applicable if the buffers are of the same size or at least of the same size groups.

**Comment by stefan2410** [comment]
@gfoidl 
Not Shared. 
    AudioPool = ArrayPool<byte>.Create( 640,100000);
    always return buffers of 1040
   Maybe you did not understand me about the allocations' cost.  I know the benchmarks and that for small buffers,. GC is more performant.  I create a cache of buffers  initially. These buffers will be re-used all  time. It should be more performant. No need to  free them, during program's life

**Comment by gfoidl** [comment]
Do you really mean `System.Buffers.ArrayPool<T>` or something different?
```c#
ArrayPool<byte> audioPool = ArrayPool<byte>.Create(640, 100000);
byte[] buffer = audioPool.Rent(1);
Debug.Assert(buffer.Length == 16);      // succeedes
```

> Maybe you did not understand me about the allocations' cost

I believe I did.

> It should be more performant.

"Should" or "is" more performant? 😉 

> this class is internal. Do you mean to copy the code?

This code is Apache License 2.0 licensed, so you can use it under these terms. Or take it as inspiration for your work.

**Comment by stefan2410** [comment]
@gfoidl  for me 
ArrayPool<byte> audioPool = ArrayPool<byte>.Create(640, 100000);   I can see in debug always 1040 .   It is .NetStandard 2.0 library as it is called from .Net 4.72
>"Should" or "is" more performant? 
Yes it is for my use case  . I know what I'm doing

**Comment by stefan2410** [comment]
@gfoidl  Hope I don't make  a foolish error, but this is what I get
ArrayPool<byte> AudioPool = ArrayPool<byte>.Create(640, 100000);
            byte[] mem = AudioPool.Rent(640);    
            //returns mem.Length = 1024

**Comment by En3Tho** [comment]
AFAIK ArrayPool is tied to power of 2, so getting 1024 is expected. For example, if you wished to get an array of 1640 bytes you would get 2048 instead.
I think if your solution is tied to that specific (640 bytes) number then you should explore custom memory manager. Also, I don't believe arraypool is suited to pool lots of same size arrays as there simply not enough space in buckets and it will resort to allocating a new array eventually.
I think @gfoidl is right and you should explore something like SlabMemoryPool

**Comment by stefan2410** [comment]
@En3Tho 
>ArrayPool is tied to power of 2, so getting 1024 is expected.

OK, but then the documentation should not write about maximum array length without to specify it . I think.  I did not know the information that you wrote.  
Yes, it is  tied to that specific (640 bytes) so I must implement my custom memory manager. 
thanks

**Comment by stefan2410** [comment]
@En3Tho
IF I accept  the extra memory cost  of the  "ArrayPool is tied to power of 2, so getting 1024 is expected." , 
Using always lots of  same size arrays .   ArrayPool.Create(640, 100000);   100,000 or 1M  in one bucket.  Allocated one time in the beginning, serving as a cache during the program life,  
then what is the problem with ArrayPool ?  I think it will be  much more peformant than the GC small allocations . And yes we have running them efficiently, although we have not full benchmarks.
It will be nice to have the SlabMemoryPool   but it is internal in the System.buffers  and  I don't prefer for the moment to create ours.

**Comment by En3Tho** [comment]
@stefan2410 I believe it always comes down to specific requirements of your task. I don't think array pool has "problems" but the generic nature of it means people did some tradeoffs when making it. Like the power of 2 thing, so it doesn't have to allocate bucket for 513, 640, 729 bytes and so on. It just places it into 1024 bucket. Also, with generic array pool you always pay the cost of calculating right array size and bucket index. I'm not sure if that's the case with custom array pool. The best way is always to measure of course. If you find your perf suitable with array pool then there is no problem at all.

**Comment by stefan2410** [comment]
@En3Tho    
We have some standard requirements (same size arrays alaways, high frequency of re-use, tradeoff with memory in order to have less CPU consumption) . So I think  ArrayPool  can do the job, till we have a custom array pool.  
What I need to know is  if I put 100,000 arrays in 1 Bucket  - the second parameter in Create, does it mean a big array , pinned in memory?
Also, I think it would be helpgul if in the documentation  of ArrayPool , there was this information 
 "ArrayPool is tied to power of 2, so getting 1024 is expected." ,

**Comment by gfoidl** [comment]
> documentation of ArrayPool

This clearly states "minimum length", the rest is an implementation detail that could be changed.
IMO the docs are clear.

> we have a custom array pool.

Copy & pasting the code from the slab-pool (and keeping the license note)?
Or you can use a `ConcurrentQueue`, etc. to put your arrays in and rent from there.
But: measure it!

> does it mean a big array , pinned in memory?

No and no.

**Comment by stefan2410** [comment]
@gfoidl
   No I don't think that documentation is OK. 
   Definetely  it has problem when  it does not mention the minimum length in power of 2.  Somebody using ArrayPool needs to know that.
> Copy & pasting the code from the slab-pool (and keeping the license note)?
No we don't do copy and pasting.  We need  battle proven tools

>does it mean a big array , pinned in memory?
>No and no.
 What it means?  that information is useful to me

**Comment by gfoidl** [comment]
> it has problem when it does not mention the minimum length in power of 2

Why should this be documented (as an implementation detail)? The [docs](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1.rent?view=net-5.0#System_Buffers_ArrayPool_1_Rent_System_Int32_) say:

> An array of type T that is at least minimumLength in length.

> What it means? that information is useful to me

https://docs.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1.create?view=net-5.0 what's unclear hear (I'm aksing to help clarify).

**Comment by stefan2410** [comment]
```
Parameters
maxArrayLength
Int32
The maximum length of an array instance that may be stored in the pool.
```
 Something like this  .   Because when I give maximum length.  I should estimate the following. It influences the total memory usage.
`  ArrayPool is tied to power of 2, so getting 1024 is expected. For example, if you wished to get an array of 1640 bytes you would get 2048 instead.`

**Comment by gfoidl** [comment]
> What a bucket means if I give 100,000 ?

A bucket is a specific size of array. If you rent e.g. size 1000, you'll get an array >= size 1000 (current implementation: 1024, next power of 2) from a bucket. If you rent size 2000, you'll get array >= 2000 (actual: 2048) from another bucket, that belongs to that size.

When these arrays are returned to the pool, they go into the specific bucket. In this example either to the bucket 1024 or to the bucket 2048.
If in a bucket are already enough array instances, so it's not kept in the pool and it will become eligible for GC.

So if you specify 100.000 for `maxArraysPerBucket` in your custom pool, that means that each bucket can hold up to 100.000 before returning the 100.001st will be discarded and GCed.

None of the arrays is pinned. They are "just" normal arrays, like `new T[sizeOfTheArrayFromTheBucket]`.

> maxArrayLength

I have the impression you mixed up "renting" from the pool and "creating a custom pool".
The maxArrayLength here refers to creating a pool, which may keep arrays up to maxArrayLength in the pool. This is nothing to do with renting.
For renting you specifiy a size you'd like to get, and you get >= that size. That this size is rounded up to the next power of 2 is an implementation detail (with potential to change, thus doesn't need to be documented, as one should not rely on that fact).

**Comment by stefan2410** [comment]
@gfoidl   I think I don't speak good English or you don't understand me .
I never said I wanted to "creating a custom pool".
What I wanted is when I  create an ArrayPool _**with maximum length 640**_  and I _**rent only 640  byte arrays**_ from the **_specific pool_**, to get arrays of 640 not 1024 (and this happens always.
I'm not sure who has mixed up his understanding
Thanks for your help.  My opinion if you update the documentation you willl help  other guys that mixed up also their impressions. Bye

**Comment by gfoidl** [comment]
> never said I wanted "creating a custom pool".

We are talking about this from the beginning:

> ArrayPool.Create( 640,100000);

That's your custom pool as opposed to the shared pool by `ArrayPool<T>.Shared`.

>  create an ArrayPool with maximum length 640

This only means that arrays with length up to 640 will be pooled (if here happens any kind of rounding I don't know from top of my head), longer ones will get GCed.
The docs state this: `may be stored in each bucket in the pool`. No word about renting.
IMO the docs are clear, but that doesn't mean that they could be improved.

Again: creating a pool and renting from a pool are different things.

**Comment by stefan2410** [comment]
First time I hear the term "custom pool"   .   It is an instance of the  pool,  not a shared.    using a class of the .Net.
`Provides a resource pool that enables reusing instances of arrays.`
If I have max length and _I rent only  arrays with such length_  I expect to get arrays with such max length.  But because it  "is tied to power of 2"  I get 1024. This is understandable, a normal behavior . When @En3Tho wrote it, I agreed immediately . But I believe you should write it  in the documentation.
Now bucket is another thing 
```
   /// <summary>The default maximum length of each array in the pool (2^20).</summary>
        private const int DefaultMaxArrayLength = 1024 * 1024;
        /// <summary>The default maximum number of arrays per bucket that are available for rent.</summary>
        private const int DefaultMaxNumberOfArraysPerBucket = 50;
           private readonly Bucket[] _buckets;
```
The Bucket class has  a 2 dimensional array   [private readonly T[]?[] _buffers;](https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Private.CoreLib/src/System/Buffers/ConfigurableArrayPool.cs#L172)
So I think this object goes to large object heap. and  the GC will not try to relocate this array?   This is my understanding and that I want.

**Comment by gfoidl** [comment]
> goes to large object heap

Why? To the large object heap (LOH) go objects which are > 85.000 bytes in size.
Other objects go to small object heap (SOH) and for the pool, as they live long enough, they will eventually be in Gen-2.

> This is my understanding and that I want.

If you want pinned arrays, there is the pinned object heap (POH) introduced in .NET 5. But above you wrote about .NET 4.7.2 so this isn't an option. So I bring the slab pool (again) into play. That's what you aim for. And for your concern:

> No we don't do copy and pasting. We need battle proven tools

I believe Kestrel is battle tested / proven enough that this works.

**Comment by stefan2410** [comment]
Then make the  slab pool   public

**Comment by gfoidl** [comment]
I believe there's no need to. Each public api has additional cost and must be kept on a high compatibilty bar.
As internal pool it can be evolved, adapted as the use cases demand.

It's already "public" in that sense that you can view the code and use it.

Slab pools are very special and not a general purpose pool, so that is why it's internal (and should remain so).

**Comment by stefan2410** [comment]
So if ArrayPool  does not do our work ( I believe it will do it,  with the tradeoff of the extra memory),
we should copy  code that MS uses in critical software like the Kestrel,  but it keeps it internal (the reason is a mystery).
So  we should take the respossibility, in case that  some developer in our organization makes a mistake and change this code, in the future. I'm not ready to agree on that, I should think about.
I don't know if "Slab pools are very special" . What I know is that we can't do our work with the .Net.  (We are forced to work with .Net 4.72, because of other MS SDK)

**Comment by stefan2410** [comment]
@stephentoub 
Why you don't make the Slab pool [PinnedBlockMemoryPool] 
(https://github.com/dotnet/aspnetcore/blob/1453c5bd2cf886f6846d2cc00b8106dbd0fdbd53/src/Shared/Buffers.MemoryPool/PinnedBlockMemoryPool.cs#L13)      public ?
It is used in Kestrel, so it is battle proven, valuable code, useful in our use-case.

**Comment by stephentoub** [comment]
> Slab pool

It's no longer a slab pool.  It's now just creating arrays of the requested size on the pinned-object heap.
